packages/shacl/lib/TermCounter.js

```diff
@@ -0,0 +1,15 @@
+export default class TermCounter {
+  /**
+   * @param {import('@rdfjs/term-map/Factory.js').TermMapFactory} env
+   */
+  constructor(env) {
+    this.termMap = env.termMap()
+  }
+
+  /**
+   * @param {import('@rdfjs/types').Term} term
+   */
+  add(term) {
+    this.termMap.set(term, (this.termMap.get(term) ?? 0) + 1)
+  }
```

Comment:
Technically, the result should have a severity but the `rdf-validate-shacl` has it correct, that it may be missing. To accommodate for that case, I would allow `null` here and add a separate counter for a total

```suggestion
  constructor(env) {
    this.total = 0
    this.termMap = env.termMap()
  }

  /**
   * @param {import('@rdfjs/types').Term | null} term
   */
  add(term) {
    this.total += 1
    if (term) {  
      this.termMap.set(term, (this.termMap.get(term) ?? 0) + 1)
    }
  }
```

Reply:
initially I allowed `null` but later on I changed my mind: handling a case with no business value is accidental complexity.
The downside is I had to move the `if` in the calling code. There, I would prefer to assert that severity is not null (and throw in case the report is not valid) but I don't know how to do it in JavaScript (Non-Null Assertion Operator is TypeScript only)